highfidelity/hifi
# (case 20005) Create HFMSerializer interface for model loaders

https://highfidelity.manuscript.com/f/cases/19302/improve-model-loading-tool-chain-project-Phase-1
Sub tickets:
https://highfidelity.fogbugz.com/f/cases/20005/Create-HFMSerializer-interface-for-model-loaders

As part of Part 1 of the Model Improvement Project, this PR streamlines the inputs/outputs to the model loading classes as seen by ModelCache, though a new universal interface called HFMSerializer (typedefd from hfm::Serializer). FBXReader and similar classes are re-named to fit this interface, for example FBXReader becomes FBXSerializer, and so on. A new file in the shared library called HifiTypes is also created, which HFMSerializer uses. This PR is a follow-up to https://github.com/highfidelity/hifi/pull/14348, which created the HFM library.

## TEST PLAN
- Run these asset browser baking tests:
   - https://highfidelity.testrail.net/index.php?/cases/view/17647
   - https://highfidelity.testrail.net/index.php?/cases/view/17645
- Run these automatic tests:
  - https://github.com/highfidelity/hifi_tests/tree/master/tests/content/entity/model/modelReaders
- In this PR, verify collision models work as expected for "Basic" and "Good" collision types
  - Note: This test is different from the one in https://github.com/highfidelity/hifi/pull/14156
  - For both "Basic" and "Good" collision types:
    - Use this model url: https://gist.githubusercontent.com/sabrina-shanman/457b40f2a66b517b0dc41196bf6ab96e/raw/bf5405a23cefe3662e30d14edf4d3386a406e128/wamConsole_hull_2_grouptest.obj
    - Create the model. In a domain with edit rights, open Create Tools. Under Create tab, select Model, paste the model url into the Model URL field, then select either "Basic" or "Good" from the Automatic Collisions dropdown. Click "Add"
    - Resize the model if needed so it is near the scale of the avatar. The model should look like a grey, chair-shaped object. Then, make the avatar attempt to stand on the seat area of the model.
  - For "Basic" collision, the avatar should fail to stand on the seat area as there is an invisible sloped barrier in the way
  - For "Good" collision, the avatar should be able to stand on the seat area.
